### PR TITLE
Add source_url to generated graphml file in PageGraph

### PR DESCRIPTION
**Pull request by ShivanKaul**
Resolves https://github.com/brave/brave-browser/issues/30604

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: third_party/blink/renderer/core/brave_page_graph/page_graph.cc [hunk 1]
```diff
@@ -1104,7 +1108,11 @@ void PageGraph::RegisterDocumentNodeCreated(blink::Document* document) {
 
   const String local_tag_name(static_cast<blink::Node*>(document)->nodeName());
   auto* dom_root = AddNode<NodeDOMRoot>(node_id, local_tag_name);
-  dom_root->SetURL(NormalizeUrl(document->Url()).GetString());
+  auto url = NormalizeUrl(document->Url());
+  dom_root->SetURL(url.GetString());
+  if (!source_url_ && url.IsValid() && url.ProtocolIsInHTTPFamily()) {
+    source_url_ = url.GetString();
+  }
```
Comment (goodov) [inline third_party/blink/renderer/core/brave_page_graph/page_graph.cc:1115]: there also `PageGraph::DidCommitLoad` which also updates `NodeDOMRoot` url.

File: third_party/blink/renderer/core/brave_page_graph/page_graph.cc [hunk 2]
```diff
@@ -997,6 +997,10 @@ String PageGraph::ToGraphML() const {
                   BAD_CAST(IsRootFrame() ? "true" : "false"));
   xmlNewTextChild(desc_container_node, nullptr, BAD_CAST "frame_id",
                   XmlUtf8String(frame_id_).get());
+  if (IsRootFrame()) {
+    xmlNewTextChild(desc_container_node, NULL, BAD_CAST "url",
+                    BAD_CAST XmlUtf8String(source_url_).get());
```
Comment (goodov) [inline third_party/blink/renderer/core/brave_page_graph/page_graph.cc:1002]: you don't need `BAD_CAST` here.

File: third_party/blink/renderer/core/brave_page_graph/page_graph.cc (hunk 1 above)
Comment (ShivanKaul) [reply third_party/blink/renderer/core/brave_page_graph/page_graph.cc:1115]: I figured we would only need one of those, so put it in `PageGraph::RegisterDocumentNodeCreated`. Do you think we need it there as well?

Comment (goodov) [reply third_party/blink/renderer/core/brave_page_graph/page_graph.cc:1115]: > I figured we would only need one of those, so put it in `PageGraph::RegisterDocumentNodeCreated`. Do you think we need it there as well?

well, if you tried it and it works fine, then I guess it's enough for this exact feature 🙂

**Review by goodov — APPROVED** [review]

**Pull request merged**